Source/Core/Core/HW/DSPHLE/UCodes/GBA.cpp

```diff
@@ -15,75 +15,41 @@ namespace HLE
 {
 void ProcessGBACrypto(u32 address)
 {
-  struct sec_params_t
-  {
-    u16 key[2];
-    u16 unk1[2];
-    u16 unk2[2];
-    u32 length;
-    u32 dest_addr;
-    u32 pad[3];
-  } sec_params;
-
-  // 32 bytes from mram addr to DRAM @ 0
-  for (int i = 0; i < 8; i++, address += 4)
-    ((u32*)&sec_params)[i] = HLEMemory_Read_U32(address);
-
-  // This is the main decrypt routine
-  u16 x11 = 0, x12 = 0, x20 = 0, x21 = 0, x22 = 0, x23 = 0;
-
-  x20 = Common::swap16(sec_params.key[0]) ^ 0x6f64;
-  x21 = Common::swap16(sec_params.key[1]) ^ 0x6573;
-
-  s16 unk2 = (s8)sec_params.unk2[0];
-  if (unk2 < 0)
-  {
-    x11 = ((~unk2 + 3) << 1) | (sec_params.unk1[0] << 4);
-  }
-  else if (unk2 == 0)
-  {
-    x11 = (sec_params.unk1[0] << 1) | 0x70;
-  }
-  else  // unk2 > 0
-  {
-    x11 = ((unk2 - 1) << 1) | (sec_params.unk1[0] << 4);
-  }
-
-  s32 rounded_sub = ((sec_params.length + 7) & ~7) - 0x200;
-  u16 size = (rounded_sub < 0) ? 0 : rounded_sub >> 3;
-
-  u32 t = (((size << 16) | 0x3f80) & 0x3f80ffff) << 1;
-  s16 t_low = (s8)(t >> 8);
-  t += (t_low & size) << 16;
-  x12 = t >> 16;
-  x11 |=
-      (size & 0x4000) >> 14;  // this would be stored in ac0.h if we weren't constrained to 32bit :)
-  t = ((x11 & 0xff) << 16) + ((x12 & 0xff) << 16) + (x12 << 8);
-
-  u16 final11 = 0, final12 = 0;
-  final11 = x11 | ((t >> 8) & 0xff00) | 0x8080;
-  final12 = x12 | 0x8080;
-
-  if ((final12 & 0x200) != 0)
-  {
-    x22 = final11 ^ 0x6f64;
-    x23 = final12 ^ 0x6573;
-  }
-  else
-  {
-    x22 = final11 ^ 0x6177;
-    x23 = final12 ^ 0x614b;
-  }
-
-  // Send the result back to mram
-  *(u32*)HLEMemory_Get_Pointer(sec_params.dest_addr) = Common::swap32((x20 << 16) | x21);
-  *(u32*)HLEMemory_Get_Pointer(sec_params.dest_addr + 4) = Common::swap32((x22 << 16) | x23);
+  u32 challenge = *reinterpret_cast<u32*>(HLEMemory_Get_Pointer(address));
+  u32 logo_color = HLEMemory_Read_U32(address + 4);
+  u32 logo_speed_32 = HLEMemory_Read_U32(address + 8);
+  u32 length = HLEMemory_Read_U32(address + 12);
+  u32 dest_addr = HLEMemory_Read_U32(address + 16);
+
+  // Unwrap key from challenge (to encrypt JoyBoot program)
+  u32 key = challenge ^ 0x6f646573;
+  *reinterpret_cast<u32*>(HLEMemory_Get_Pointer(dest_addr)) = Common::swap32(key);
+
+  // Assemble authentication code (first reply to GBA)
+  u16 color_speed_coded;
+  s16 logo_speed = s8(logo_speed_32);
+  if (logo_speed < 0)
```

Comment:
This is fairly close to what a compiler would produce, would you mind rewriting those to convey the actual meaning better? Like `logo_color << 1` -> `logo_color * 2` etc.
It would also benefit from a few comments on why the values are used like that (3 different strategies to move the thing around?)